logic/odmr_logic.py

```diff
@@ -454,41 +496,69 @@ def mw_sweep_on(self):
 
-        elif self.mw_scanmode == MicrowaveMode.SWEEP:
-            if np.abs(self.mw_stop - self.mw_start) / self.mw_step >= limits.sweep_maxentries:
-                self.log.warning('Number of frequency steps too large for microwave device. '
-                                 'Lowering resolution to fit the maximum length.')
-                self.mw_step = np.abs(self.mw_stop - self.mw_start) / (limits.list_maxentries - 1)
-                self.sigParameterUpdated.emit({'mw_step': self.mw_step})
 
-            sweep_return = self._mw_device.set_sweep(
-                self.mw_start, self.mw_stop, self.mw_step, self.sweep_mw_power)
-            self.mw_start, self.mw_stop, self.mw_step, self.sweep_mw_power, mode = sweep_return
+                # adjust the end frequency in order to have an integer multiple of step size
+                # The master module (i.e. GUI) will be notified about the changed end frequency
+                final_freq_list.extend(freq_list)
+                # self.frequency_lists.append(freq_list)
+
+                used_starts.append(mw_start)
+                used_steps.append(mw_step)
+                used_stops.append(end_freq)
 
-            param_dict = {'mw_start': self.mw_start, 'mw_stop': self.mw_stop,
-                          'mw_step': self.mw_step, 'sweep_mw_power': self.sweep_mw_power}
+
+
+            freq_list, self.sweep_mw_power, mode = self._mw_device.set_list(final_freq_list,
+                                                                            self.sweep_mw_power)
+
+            self.final_freq_list = freq_list
+            self.mw_starts = used_starts
+            self.mw_stops = used_stops
+            self.mw_steps = used_steps
+            param_dict = {'mw_starts': used_starts, 'mw_stops': used_stops,
+                          'mw_steps': used_steps, 'sweep_mw_power': self.sweep_mw_power}
+
+            self.sigParameterUpdated.emit(param_dict)
+
+        elif self.mw_scanmode == MicrowaveMode.SWEEP:
+            if self.ranges == 1:
+                mw_stop = self.mw_stops[0]
+                mw_step = self.mw_steps[0]
+                mw_start = self.mw_starts[0]
+
+                if np.abs(mw_stop - mw_start) / mw_step >= limits.sweep_maxentries:
+                    self.log.warning('Number of frequency steps too large for microwave device. '
+                                     'Lowering resolution to fit the maximum length.')
+                    mw_step = np.abs(mw_stop - mw_start) / (limits.list_maxentries - 1)
+                    self.sigParameterUpdated.emit({'mw_steps': [mw_step]})
+
+                sweep_return = self._mw_device.set_sweep(
+                    mw_start, mw_stop, mw_step, self.sweep_mw_power)
+                mw_start, mw_stop, mw_step, self.sweep_mw_power, mode = sweep_return
+
+                param_dict = {'mw_starts': [mw_start], 'mw_stops': [mw_stop],
+                              'mw_steps': [mw_step], 'sweep_mw_power': self.sweep_mw_power}
```

Comment:
1) The data for the lists self.frequency_lists and self.freq_final_list is not set here. I think this causes the error when working in SWEEP mode.